frontend/src/main/scala/bloop/engine/tasks/ScalaNativeToolchain.scala

```diff
@@ -0,0 +1,127 @@
+
+  /**
+   * Compile down to native binary using Scala Native's toolchain.
+   *
+   * @param project  The project to link
+   * @param entry    The fully qualified main class name
+   * @param logger   The logger to use
+   * @param optimize The configuration of the optimizer.
+   * @return The absolute path to the native binary.
+   */
+  def link(project: Project,
+           entry: String,
+           logger: Logger,
+           optimize: OptimizerConfig): Task[Try[AbsolutePath]] = {
+
+    val bridgeClazz = classLoader.loadClass("bloop.scalanative.NativeBridge")
+    val paramTypes = classOf[Project] :: classOf[String] :: classOf[Logger] :: classOf[
+      OptimizerConfig] :: Nil
+    val nativeLinkMeth = bridgeClazz.getMethod("nativeLink", paramTypes: _*)
+
+    // The Scala Native toolchain expects to receive the module class' name
+    val fullEntry = if (entry.endsWith("$")) entry else entry + "$"
+
+    Task(nativeLinkMeth.invoke(null, project, fullEntry, logger, optimize)).materialize.map {
+      _.collect { case path: Path => AbsolutePath(path) }
+    }
+  }
+
+  /**
+   * Link `project` to a native binary and run it.
+   *
+   * @param state    The current state of Bloop.
+   * @param project  The project to link.
+   * @param cwd      The working directory in which to start the process.
+   * @param main     The fully qualified main class name.
+   * @param args     The arguments to pass to the program.
+   * @param optimize The configuration of the optimizer.
+   * @return A task that links and run the project.
+   */
+  def run(state: State,
+          project: Project,
+          cwd: AbsolutePath,
+          main: String,
+          args: Array[String],
+          optimize: OptimizerConfig): Task[State] = {
+    link(project, main, state.logger, optimize).flatMap {
+      case Success(nativeBinary) =>
+        val cmd = nativeBinary.syntax +: args
+        Forker.run(cwd, cmd, state.logger, state.commonOptions).map { exitCode =>
+          val exitStatus = {
+            if (exitCode == Forker.EXIT_OK) ExitStatus.Ok
+            else ExitStatus.RunError
+          }
+          state.mergeStatus(exitStatus)
+        }
+      case Failure(ex) =>
+        Task {
+          state.logger.error("Couldn't create native binary.")
+          state.logger.trace(ex)
+          state.mergeStatus(ExitStatus.LinkingError)
+        }
+    }
+  }
+
+}
+
+object ScalaNativeToolchain {
+
+  private[this] val instancesCache: ConcurrentHashMap[Array[AbsolutePath], ScalaNativeToolchain] =
+    new ConcurrentHashMap
+
+  def forProject(project: Project, logger: Logger): ScalaNativeToolchain = {
+    project.nativeConfig match {
+      case None =>
+        resolveNativeToolchain(logger)
+      case Some(config) =>
+        val classpath = config.toolchainClasspath.map(AbsolutePath.apply)
+        direct(classpath)
+    }
+  }
+
+  def resolveNativeToolchain(logger: Logger): ScalaNativeToolchain = {
+    val jars = bridgeJars(logger)
```

Comment:
Last time I was touching `ScalaInstance` I noticed that the overhead of calling coursier is high, and skipping its invocation reduced starting up time significantly.

Reply:
I think we should do the same we do with ScalaInstance here https://github.com/scalacenter/bloop/blob/topic/allow-only-java-projects/backend/src/main/scala/bloop/ScalaInstance.scala#L106-L107 so that we can share the resolution of the bridges across every project not defining the `toolchainClasspath`.

Reply:
How high was it in your measurement? Resolving `bloop-native-bridge_2.12` is pretty fast on my machine, but I may very well be benchmarking it wrong:

```
[info] Benchmark                                                Mode   Cnt   Score   Error  Units
[info] Bench.resolveNativeBridge                              sample  4056  24.671 ± 0.086  ms/op
[info] Bench.resolveNativeBridge:resolveNativeBridge·p0.00    sample        22.315          ms/op
[info] Bench.resolveNativeBridge:resolveNativeBridge·p0.50    sample        24.052          ms/op
[info] Bench.resolveNativeBridge:resolveNativeBridge·p0.90    sample        27.197          ms/op
[info] Bench.resolveNativeBridge:resolveNativeBridge·p0.95    sample        27.787          ms/op
[info] Bench.resolveNativeBridge:resolveNativeBridge·p0.99    sample        29.014          ms/op
[info] Bench.resolveNativeBridge:resolveNativeBridge·p0.999   sample        30.079          ms/op
[info] Bench.resolveNativeBridge:resolveNativeBridge·p0.9999  sample        31.818          ms/op
[info] Bench.resolveNativeBridge:resolveNativeBridge·p1.00    sample        31.818          ms/op
```

The resolution of bridges was shared between projects that don't define `toolchainClasspath`. Should I revert https://github.com/scalacenter/bloop/pull/457/commits/27cd1a32bdd1d2d7c18acc7bce155bdbbe0fdf28 ?